### PR TITLE
Optimize data loader and eliminate redundant feature value tensors

### PR DESCRIPTION
**Pull request by TonyCongqianWang**
# Eliminate Redundant Feature Value Tensors & Optimize H2D Pipeline

## Overview
This PR eliminates the redundant `white_values` and `black_values` float tensors (which always hold `1.0f` for active features and `0.0f` for empty slots) from the entire pipeline. 

By replacing explicit value transfers and lookups with implicit masks/assumptions, we reduce Host-to-Device (H2D) transfer sizes, shrink memory allocations, and simplify both the C++ dataloader and GPU kernels.

### 1. C++ Dataloader & Python Bindings
* **Value Tensor Elimination**: Removed `white_values` and `black_values` arrays from `SparseBatch` structure.
* **Heap Size Reduction**: Slashed contiguous `m_float_block` size to `size * 3` floats (down from `size * 3 + size * max_active * 2`).
* **Piece Count Packing (Optimization)**: Replaced `psqt_indices` and `layer_stack_indices` arrays with a single raw `piece_count` array. `psqt_indices` and `layer_stack_indices` are now calculated dynamically on the GPU in PyTorch (`data_loader/_native.py`):
  ```python
  psqt_indices = (piece_count - 1) // 4
  layer_stack_indices = psqt_indices
  ```
  This reduces integer PCIe H2D transfers and isolates the bucketing logic to Python.

### 2. CUDA Kernels & Autograd
* **Cupy Kernels**: Removed `input_values` from signatures of both `sparse_input_linear_forward` and `sparse_input_linear_backward`. Accumulations are performed directly without scaling by the input values:
  * **Forward**: `shared_output_slice[s] += weight_slice[s];`
  * **Backward**: `atomicAdd(&weight_grad_slice[s], sog);`
* **Autograd & Fallback**: Updated CPU/MPS fallback `_torch_sparse_linear` and `_CudaSparseLinearFunction` to drop `feature_values` and compute weight scaling implicitly using `mask.to(weight.dtype)`.

### 3. Downstream Training, Evaluation, and Tests
* Updated `LightningModule`, `cross_check_eval.py`, `ftperm.py`, and `perf_sigmoid_fitter.py` to unpack exactly 8 tensors (down from 10) and call the model without value tensors.
* Updated unit tests (`test_feature_transformer.py`, `test_sparse_linear_device_parity.py`) to align with the new signatures.

### CODE REVIEW

**Comment by TonyCongqianWang** [comment]
Most of the data (in terms of memory) we transfer to GPU is actually completely useless as white_values and black_values are always binary and completely determined by active_indices.

Unfortunately it is never clear whether it makes a difference in practice or not. As such I would greatly appreciate you testing the speed

@vondele @Disservin 

This PR superseeds #482

**Comment by TonyCongqianWang** [comment]
From my local testing, the data_loader alone is almost 50% faster! This can be somewhat massive overall

**Comment by vondele** [comment]
Locally, nice speedup on the dataloader:
```
                        master          1 Iter:    18000   Time(s):     56.4   MPos/s:     20.9   It/s:      319   MB/s: 8.35e+03   B/pos:      418
                       fb6eaa5          1 Iter:    18000   Time(s):     42.5   MPos/s:     27.8   It/s:      424   MB/s: 1.05e+04   B/pos:      394
                       cff6dc5          1 Iter:    18000   Time(s):     60.5   MPos/s:     19.5   It/s:      297   MB/s: 7.85e+03   B/pos:      422
                       6882c0b          1 Iter:    18000   Time(s):     59.4   MPos/s:     19.8   It/s:      303   MB/s: 7.98e+03   B/pos:      421
                        master          2 Iter:    18000   Time(s):     55.6   MPos/s:     21.2   It/s:      324   MB/s: 8.46e+03   B/pos:      417
                       fb6eaa5          2 Iter:    18000   Time(s):     42.4   MPos/s:     27.8   It/s:      424   MB/s: 1.05e+04   B/pos:      394
                       cff6dc5          2 Iter:    18000   Time(s):     60.3   MPos/s:     19.5   It/s:      298   MB/s: 7.84e+03   B/pos:      420
                       6882c0b          2 Iter:    18000   Time(s):     59.7   MPos/s:     19.8   It/s:      301   MB/s: 7.95e+03   B/pos:      422
                        master          3 Iter:    18000   Time(s):     56.1   MPos/s:       21   It/s:      321   MB/s: 8.36e+03   B/pos:      416
                       fb6eaa5          3 Iter:    18000   Time(s):     43.1   MPos/s:     27.3   It/s:      417   MB/s: 1.04e+04   B/pos:      399
```
Also includes testing for https://github.com/official-stockfish/nnue-pytorch/pull/481 (no speedup)

**Comment by TonyCongqianWang** [comment]
Very nice speedup indeed. It should also translate to training hopefully.

I'll make a nettest PR later

**Comment by vondele** [comment]
So, actual training speeds
```
==== checkout master =====
==== current sha: 89d572575441c927947fdc8b98c41719c530d120 ====
==== compile data loader ====
==== train with extra opts:  ====
master 1 65.59it/s
master 2 68.13it/s
master 3 68.52it/s
==== checkout remove-redundant-feature-values =====
==== current sha: fb6eaa59f09a02cdcd07dd2786990b2ba5190cb3 ====
==== compile data loader ====
==== train with extra opts:  ====
remove-redundant-feature-values 1 70.45it/s
remove-redundant-feature-values 2 70.36it/s
remove-redundant-feature-values 3 70.69it/s
==== checkout optimize-parallel-data-reader =====
==== current sha: e9de03366e93efe17cdda8ae216ab965cd599617 ====
==== compile data loader ====
==== train with extra opts:  ====
optimize-parallel-data-reader 1 67.76it/s
optimize-parallel-data-reader 2 68.34it/s
optimize-parallel-data-reader 3 67.91it/s
```

Includes testing for https://github.com/official-stockfish/nnue-pytorch/pull/481

**Comment by TonyCongqianWang** [comment]
Great. Shall we make a small test run to verify nothing breaks and merge? Maybe @Disservin, can you also test if you get a speedup or at least no slowdown ?

**Comment by Disservin** [comment]
```
master (the other day)
Epoch 129 (Train):  60%|   915/1526  [00:31<00:20,  29.35it/s, train_loss=0.00393,  v_num=1]
master (today):
Epoch 129 (Train): 100%|  1526/1526  [00:53<00:00,  28.62it/s, train_loss=0.00368,  v_num=0]
pr:
Epoch 129 (Train): 100%|  1526/1526  [00:49<00:00,  30.83it/s, train_loss=0.00386,  v_num=0]
```

**Comment by TonyCongqianWang** [comment]
Great! Speedup isn't as large as I'd hoped, but it's something at least

**Comment by TonyCongqianWang** [comment]
Passed nettest.

https://github.com/vondele/nettest/pull/363

Ready to merge?

**Comment by xu-shawn** [comment]
If I understand correctly, this PR packs three changes at once:
1. Removing feature value tensors
2. Coalescing the rest of tensors into int and float blocks
3. Calculate buckets on the fly

1 is a free speedup and should be merged right away. But are we sure that 2+3 contributes to the benchmarking gains? Maybe we should split this into 2 PRs and test the second part in isolation

**Comment by TonyCongqianWang** [comment]
> But are we sure that 2 contributes to the benchmarking gains? Maybe we should split this into 2 PRs and test the second part in isolation


We already tested that part in isolation!

It used to be only part 2 and vondele tested it to be a very slight training speedup and a somewhat decent data_loader_bench speedup on his machine.

After adding part 1 the speedup did increase on both fronts. Again only very slightly on the actual training.

We haven't run coderabbit or copilot yet I think, so we can use that to review before meeging

File: data_loader/_native.py [hunk 1]
```diff
@@ -27,77 +42,54 @@ class SparseBatch(ctypes.Structure):
...
+        us = float_block_gpu[0 : size].view(size, 1)
+        outcome = float_block_gpu[size : 2 * size].view(size, 1)
+        score = float_block_gpu[2 * size : 3 * size].view(size, 1)
+
+        white_indices = int_block_gpu[0 : size * max_active].view(size, max_active)
+        black_indices = int_block_gpu[size * max_active : 2 * size * max_active].view(size, max_active)
+        piece_count = int_block_gpu[2 * size * max_active : 2 * size * max_active + size].view(size).long()
+        
+        # Calculate indices based on the piece count on the target device
+        psqt_indices = (piece_count - 1) // 4
```
Comment (xu-shawn) [inline data_loader/_native.py]: I don't like that this architecture-specific calculation is now inside the python dataloader wrapper. Can we move this into `NNUEModel.forward`?

Comment (TonyCongqianWang) [reply data_loader/_native.py]: Sure..

Comment (TonyCongqianWang) [reply data_loader/_native.py]: Though that was always in the dataloader lol -- which is honestly also something I disliked

**Comment by TonyCongqianWang** [comment]
So whats the state on this?

@vondele @Disservin @xu-shawn  Anyone want to run copilot or code rabbit? I can change the signatures but I'd rather do all requested changes at once so that we can merge afterwards

File: data_loader/_native.py [hunk 2]
```diff
@@ -27,77 +42,54 @@ class SparseBatch(ctypes.Structure):
...
+        layer_stack_indices = psqt_indices
```
[the remark this answers is not in the document]
Comment (TonyCongqianWang) [reply data_loader/_native.py]: I feel like there is some truth, but some logic was missed:

if device is CPU, there is no pinning at all. Thus the returned batch isn't "partially pinned".

**Comment by Disservin** [comment]
merged in #490, #491, #492
